File: app/src/main/java/com/moroom/android/navui/profile/ProfileFragment.java
package com.moroom.android.navui.profile;

import android.app.AlertDialog;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;

import com.moroom.android.LoginActivity;
import com.moroom.android.MyReviewActivity;
import com.moroom.android.R;
import com.moroom.android.databinding.FragmentProfileBinding;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.FirebaseDatabase;

public class ProfileFragment extends Fragment {
    private FragmentProfileBinding binding;
    private final FirebaseDatabase database = FirebaseDatabase.getInstance();
    private final FirebaseUser user = FirebaseAuth.getInstance().getCurrentUser();

    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentProfileBinding.inflate(inflater, container, false);
        View view = binding.getRoot();

        setupView();
        setupListener();

        return view;
    }

    private void setupView() {
        if (user != null) showUserProfile();
        else goToLogin();
    }

    private void showUserProfile() {
        // 로그인 된 유저의 `프로필`을 보여줌
        binding.tvUserEmail.setText(user.getEmail());
    }

    private void goToLogin() {
        binding.moveToLogin.setVisibility(View.VISIBLE);

        binding.tvUserEmail.setVisibility(View.GONE);
        binding.tvMyReview.setVisibility(View.GONE);
        binding.tvLogout.setVisibility(View.GONE);
        binding.tvDeleteAccount.setVisibility(View.GONE);
        binding.tvMoRoomVersion.setVisibility(View.GONE);
        binding.linearTop.setVisibility(View.GONE);
        binding.linearSeparation1.setVisibility(View.GONE);
        binding.linearSeparation2.setVisibility(View.GONE);
        binding.linearSeparation3.setVisibility(View.GONE);
        binding.linearSeparation4.setVisibility(View.GONE);
    }

    private void setupListener() {
        binding.tvMyReview.setOnClickListener(v -> {
            Intent intent = new Intent(getActivity(), MyReviewActivity.class);
            startActivity(intent);
        });

        binding.tvLogout.setOnClickListener(v -> showLogoutDialog());

        binding.tvDeleteAccount.setOnClickListener(v -> showDeleteDialog());

        binding.moveToLogin.setOnClickListener(v -> {
            Intent intent = new Intent(getActivity(), LoginActivity.class);
            startActivity(intent);
            getActivity().finish();
        });
    }

    private void showLogoutDialog() {
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        builder.setTitle(getString(R.string.log_out));
        builder.setMessage(getString(R.string.log_out_box));
        builder.setPositiveButton(getString(R.string.yes), (dialogInterface, i) -> logout());
        builder.setNegativeButton(getString(R.string.no), null);
        AlertDialog dialog = builder.create();
        dialog.show();
    }

    private void showDeleteDialog() {
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        builder.setTitle(getString(R.string.delete_account));
        builder.setMessage(getString(R.string.delete_account_box));
        builder.setPositiveButton(getString(R.string.yes), (dialogInterface, i) -> deleteAccount());
        builder.setNegativeButton(getString(R.string.no), null);
        AlertDialog dialog = builder.create();
        dialog.show();
    }

    private void logout() {
        FirebaseAuth.getInstance().signOut();
        Intent intent = new Intent(getActivity(), LoginActivity.class);
        startActivity(intent);
        getActivity().finish();
    }

    private void deleteAccount() {
        // delete Authentication data
        user.delete()
                .addOnCompleteListener(task -> {
                    if (task.isSuccessful()) {
                        Intent intent = new Intent(getActivity(), LoginActivity.class);
                        startActivity(intent);
                        Toast.makeText(getActivity(), getString(R.string.delete_complete), Toast.LENGTH_SHORT).show();
                    }
                });

        database.getReference("UserAccount").child(user.getUid()).removeValue();
        database.getReference("Address").child(user.getUid()).getParent().removeValue();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
}

File: app/src/main/java/com/moroom/android/presentation/nav/profile/ProfileFragment.kt
package com.moroom.android.presentation.nav.profile

import android.app.AlertDialog
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import com.google.firebase.auth.FirebaseUser
import com.moroom.android.R
import com.moroom.android.databinding.FragmentProfileBinding
import com.moroom.android.presentation.login.LoginActivity
import com.moroom.android.presentation.nav.MainActivity
import dagger.hilt.android.AndroidEntryPoint

@AndroidEntryPoint
class ProfileFragment : Fragment() {
    private var _binding: FragmentProfileBinding? = null
    private val binding get() = _binding!!
    private val viewModel: ProfileViewModel by viewModels()

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentProfileBinding.inflate(inflater, container, false)
        val view: View = binding.root
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setUpView()
        setUpObserver()
        setUpListener()
    }

    private fun setUpView() {
        val currentUser = viewModel.currentUser
        if (currentUser != null) showUserProfile(currentUser)
        else showGoToLogin()
    }

    private fun showUserProfile(currentUser: FirebaseUser) {
        binding.tvUserEmail.text = currentUser.email
    }

    private fun setUpObserver() {
        viewModel.isUserDeleted.observe(viewLifecycleOwner) {
            if (it) {
                Toast.makeText(activity, getString(R.string.delete_complete), Toast.LENGTH_SHORT)
                    .show()
                navigateToMain()
            } else {
                Toast.makeText(activity, getString(R.string.user_delete_error), Toast.LENGTH_LONG)
                    .show()
            }
        }
    }

    private fun setUpListener() {
        binding.tvLogout.setOnClickListener { showLogoutDialog() }
        binding.tvDeleteAccount.setOnClickListener { showDeleteDialog() }
        binding.moveToLogin.setOnClickListener { navigateToLogin() }
    }

    private fun showLogoutDialog() {
        val builder = AlertDialog.Builder(requireContext())
        val dialog = builder.create()
        builder.setTitle(getString(R.string.log_out))
        builder.setMessage(getString(R.string.log_out_box))
        builder.setPositiveButton(getString(R.string.yes)) { _: DialogInterface?, _: Int -> viewModel.logout(); navigateToMain() }
        builder.setNegativeButton(getString(R.string.no), null)
        dialog.show()
    }

    private fun showDeleteDialog() {
        val builder = AlertDialog.Builder(requireContext())
        val dialog = builder.create()
        builder.setTitle(getString(R.string.delete_account))
        builder.setMessage(getString(R.string.delete_account_box))
        builder.setPositiveButton(getString(R.string.yes)) { _: DialogInterface?, _: Int -> viewModel.deleteUserData() }
        builder.setNegativeButton(getString(R.string.no), null)
        dialog.show()
    }

    private fun showGoToLogin() {
        binding.apply {
            tvUserEmail.visibility = View.GONE
            tvLogout.visibility = View.GONE
            tvDeleteAccount.visibility = View.GONE
            tvMoRoomVersion.visibility = View.GONE
            linearTop.visibility = View.GONE
            linearSeparation3.visibility = View.GONE
            linearSeparation4.visibility = View.GONE
            linearSeparation1.visibility = View.GONE
            linearSeparation2.visibility = View.GONE

            moveToLogin.visibility = View.VISIBLE
        }
    }

    private fun navigateToLogin() {
        val intent = Intent(activity, LoginActivity::class.java)
        startActivity(intent)
    }

    private fun navigateToMain() {
        val intent = Intent(activity, MainActivity::class.java)
        startActivity(intent)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}
